src/main/ngapp/accio/src/app/mask.directive.ts

```diff
@@ -0,0 +1,142 @@
+    this.drawMask(maskPixels, this.scaledData, 1);
+  }
+
+  /**
+   *  Creates a UI mask to the user by reading in all values in the given maskPixels
+   *  Clears the canvas fisrt so the mask can be saved as a new image and then over-
+   *    lays the mask on top of the scaled image. 
+   *  Adds all pixels to mask by giving alpha value of 255, changing global alpha on 
+   *    maskImage will multiply 255 by globalAlpha = [0,1]
+   *  @Param: maskPixels: set of coordinates indexed by red value of pixels to be in mask
+   *  @Param: imgData: Uint8clampedArray of all pixel data in scaled image. 
+   *  @Param: alphaValue: the number the user passes in on a slide bar for transparency of the mask
+   */
+  private drawMask(maskPixels: Set<number>, scaledData: ImageData, alphaValue: number): void {
+    //clear old context
+    this.ctx.clearRect(0,0,this.canvas.nativeElement.width, this.canvas.nativeElement.height);
+
+    //  reset global alpha if image was tinted
+    this.ctx.globalAlpha = 1;
+
+    // access all pixels in original mask, scale them up to image on UI width
+    for (let pixel of maskPixels) {
+      let y = this.getY(this.originalData.width, pixel);
+      let x = this.getX(this.originalData.width, pixel, y);
+      this.maskData.data[this.getPixel(x, y, scaledData.width) + 3] = 255;
+    } 
+    // Put the new img data on ctx and save as an img
+    this.ctx.putImageData(this.maskData, 0, 0);
+    
+    const maskUrl = this.canvas.nativeElement.toDataURL();
+
+    // clear canvas again to prepare for image and mask
+    this.ctx.clearRect(0,0,this.canvas.nativeElement.width, this.canvas.nativeElement.height);
+  
+    console.log('adding mask'); 
+    this.drawImageAndMask(scaledData, maskUrl, alphaValue)
+    console.log('new mask added');
+  }
+
+  /**
+  * TODO: alpha change causes old mask to get darker as new pixels added
+  * Changes the alpha value of the mask displayed 
+  * @Param: alphaValue: the number the user passes in on a slide bar for transparency of the mask
+  */
+  private changeAlpha(alphaValue: number): void {
+    if (alphaValue == 0) {
+      alphaValue = .01;
+    }
+    this.ctx.globalAlpha = alphaValue;
+    var newMask = new Image();
+    newMask.src = this.maskImage.src;
+    newMask.onload = () => {
+        this.ctx.drawImage(newMask, 0, 0);
+    };
+  }
+
+  /**
+  * draws the original image and updated mask on context
+  * @Param: imgData: ImageData from the original image
+  * @Param: mask: source of new mask 
+  * @Param: alphaValue: the number the user passes in on a slide bar for transparency of the mask
+  */
+  private drawImageAndMask(scaledData: ImageData, maskUrl: string, alphaValue: number): void {
+    // restore original image (scaledData)
+    this.ctx.putImageData(scaledData, 0, 0);
+
+    //make a new image to add
+    let maskImage = new Image();
+
+    //TODO: alpha change causes old mask to get darker as new pixels added
+    //this.ctx.globalAlpha = alphaValue;
```

Comment:
One possible explanation is if you get the mask pixels from the canvas over time.  This would most likely incorporate the globalAlpha which would explain why adding more pixels the older pixels get darker and darker.

Reply:
But if I reset the global alpha to 1 before I get that pixels and then reset them all to the alpha value shouldn't this fix the issue of them getting darker and darker because they all start at the same value?